### ISSUE TITLE
Single Fetch throwing response does not work

### ISSUE THREAD

**Issue by timvandam** [remix-run/remix#9793]
### Reproduction

https://stackblitz.com/edit/remix-run-remix-ndbnrb?file=app%2Froutes%2F_index.tsx,app%2Froot.tsx,tsconfig.json

### System Info

```shell
System:
    OS: macOS 14.3
    CPU: (16) arm64 Apple M3 Max
    Memory: 229.22 MB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.11.1 - ~/.nvm/versions/node/v20.11.1/bin/node
    npm: 10.2.4 - ~/.nvm/versions/node/v20.11.1/bin/npm
    pnpm: 9.5.0 - ~/.nvm/versions/node/v20.11.1/bin/pnpm
  Browsers:
    Chrome: 127.0.6533.72
    Safari: 17.3
```

### Used Package Manager

npm

### Expected Behavior

Loader states should always stay consistent. E.g. `useLoaderData<typeof loader>` should always contain the latest valid loader data

### Actual Behavior

Loader state isnt consistent when using `throw response` with single fetch. It becomes undefined unexpectedly leading to destructuring errors (see stackblitz)

**Comment by timvandam** [comment]
@brophdawg11 Thanks for your explanation, however, I am throwing the response stub which per the docs is a way to the short circuit actions/loaders to immediately respond to the incoming request. So what is happening here is that an error is incorrectly thrown by Remix, while it should have just sent a normal response

**Comment by brophdawg11** [comment]
>  So what is happening here is that an error is incorrectly thrown by Remix

How so?  Your code is throwing the response stub - and any time you `throw` _anything_ from a `loader`/`action` you trigger an `ErrorBoundary`.

Can you point to the section of the docs that is causing the confusion?

**Comment by brophdawg11** [comment]
Redirects are different than errors since you're telling the router to navigate to a new route, as opposed to informing it of an error on the current route.

You're not throwing a `redirect` in your reproduction though - you're throwing a response stub with a 400 status.  Without single fetch, that's equivalent to `throw json({ error: "broken" }, { status: 400 })` which would trigger the `ErrorBoundary` in the same way.

**Comment by timvandam** [comment]
> Redirects are different than errors since you're telling the router to navigate to a new route, as opposed to informing it of an error on the current route.
> 
> You're not throwing a `redirect` in your reproduction though - you're throwing a response stub with a 400 status. Without single fetch, that's equivalent to `throw json({ error: "broken" }, { status: 400 })` which would trigger the `ErrorBoundary` in the same way.

You're right, I've only ever thrown with redirect before so didn't realize it worked this way. I notice though that single fetch's types override `useRouteLoaderData` always return something, i.e., it removes the `| undefined` from the return type (see https://github.com/remix-run/remix/blob/ac747ff7cdd8dde2b50fdf392fce0a29bbf2257f/packages/remix-react/future/single-fetch.d.ts#L27-L29). This seems inconsistent with this behavior

**Comment by brophdawg11** [comment]
> it removes the | undefined from the return type

Ah thanks for pointing this out - this is a bug.  In your case `useRouteLoaderData("root")` will always be defined but if you put an invalid route id or a non-mounted id in there it would be `undefined` - got this fixed in #9796 👍

**Comment by timvandam** [comment]
How exactly is the behavior different between the two, and for what reason? It’s not clear from the docs. I only seitched to it for the | ubdefined and was surprised it never became undefined during errors unlike useLoaderData
